🎸 🧨 Add frequency baselines

mberr opened the pull request:
This PR adds a simple, non-parametric baseline approach based upon entity-relation co-occurrence. While being inferior to properly tuned models, it achieves surprisingly strong results (23% MRR / 17% H@1 / 35% H@10) on FB15k-237. It does not require any training (although there is an indexing phase for faster evaluation).

From a library design perspective, it is interesting since it only fullfils what is necessary for evaluation (`score_h` / `score_t`), but does not fully implement the `Model` interface.

mberr committed:
Add pseudo-typing / entity-relation co-occurrence baseline

mberr committed:
Add optional normalization

mberr committed:
Add significantly worse entity co-occurrence baseline (ignoring relations)

mberr committed:
Reduce code duplication

cthoyt committed:
Update rank_based_evaluator.py

cthoyt committed:
Generalize to all datasets

cthoyt committed:
Run on multiple models

cthoyt committed:
Update baseline.py

cthoyt committed:
Reorganize and update logging

mberr committed:
Add inverse triples baseline

mberr committed:
Add inverse triples baseline to models

mberr committed:
fix shape

mberr committed:
fix main loop

cthoyt committed:
Add time and count statistics

cthoyt committed:
Show kwarg values in result table

cthoyt committed:
Add missing applications of normalize

cthoyt committed:
Update baseline.py

cthoyt committed:
Rename columns

mberr committed:
some sparse matrix optimizations

mberr committed:
use threshold by default

cthoyt committed:
Update baseline.py

cthoyt committed:
Add plotting

cthoyt committed:
Pass flake8

cthoyt committed:
Pass mypy

cthoyt committed:
Make row indices explicit

cthoyt committed:
Merge branch 'master' into frequency-base-lines

mberr committed:
Merge remote-tracking branch 'origin/frequency-base-lines' into frequency-base-lines

# Conflicts:
#	src/pykeen/models/baseline.py

cthoyt committed:
Run multiple trials and update plots

cthoyt committed:
MP4eva

cthoyt committed:
Update plotting

cthoyt committed:
Add test version of benchmark

Trigger CI

cthoyt committed:
Update baseline.py

Trigger CI

mberr committed:
Update documentation of PseudoTypeBaseline

cthoyt committed:
Simplify imports and CLI

mberr committed:
Merge baselines

mberr committed:
add unittests

mberr committed:
use float32

mberr committed:
Add base case and swap default

mberr committed:
reduce code duplication

mberr committed:
update docstring

mberr committed:
Merge remote-tracking branch 'origin/frequency-base-lines' into frequency-base-lines

mberr committed:
remove dead code

mberr committed:
Add docstring

mberr committed:
Add todo

cthoyt committed:
Update docs

cthoyt committed:
Update tests

cthoyt committed:
Update baseline.py

mberr commented on src/pykeen/models/baseline.py:
```diff
@@ -54,21 +60,32 @@ def get_csr_matrix(
+        """
+        super().__init__(
+            triples_factory=triples_factory,
+            # These operations are deterministic and a random feed can be fixed
```
```suggestion
            # These operations are deterministic and a random seed can be fixed
```

mberr commented on src/pykeen/models/baseline.py:
```diff
@@ -0,0 +1,464 @@
+        )
+
+
+# TODO: Remove this model from this PR?
```
@cthoyt any comments on this? I think it might be worth to split this PR into two parts, where one only adds the `MarginalDistributionBaseline`, and keeps out the inverse triple one for now.

For the inverse triple one there is still some work to be done for reducing code duplication with the leakage detection.

cthoyt replied:
I'd be okay with splitting the PR so we have something done and merged first

cthoyt committed:
Update baseline.py

mberr committed:
fix probabilities?

mberr committed:
Merge remote-tracking branch 'origin/frequency-base-lines' into frequency-base-lines

mberr committed:
Add TODO

cthoyt committed:
Fix imports

cthoyt committed:
Add citation and update readme

mberr committed:
add hotfix

mberr committed:
Merge remote-tracking branch 'origin/frequency-base-lines' into frequency-base-lines

cthoyt committed:
Merge branch 'master' into frequency-base-lines

cthoyt committed:
asdsad

cthoyt committed:
Update src/pykeen/models/baseline.py

Co-authored-by: Max Berrendorf <berrendorf@dbs.ifi.lmu.de>

cthoyt committed:
Reorganize code

cthoyt committed:
Update test_models.py

cthoyt committed:
Move test code

cthoyt commented on src/pykeen/models/baseline/models.py:
```diff
@@ -0,0 +1,184 @@
+    .. math ::
+        P(t | h, r) = P(t | h) * P(t | r)
+
+    Depending on the settings, we either set $P(t | *) = \frac{1}{n}$, or estimate them by counting occurrences in the
```
what is `n`? Number of entities?

mberr replied:
Yes. We this model uses entity scores = probability distribution over all entities.

cthoyt replied:
Updated text in bee29d47

cthoyt commented on src/pykeen/models/baseline/models.py:
```diff
@@ -0,0 +1,184 @@
+    def __init__(
+        self,
+        triples_factory: CoreTriplesFactory,
+        entity_margin: bool = True,
```
what happens if `entity_margin` and `relation_margin` are both false? Should that raise an exception?

mberr replied:
I first has this exception, but then reverted it again. It will lead to a uniform distribution, i.e. equal scores for all entities.

cthoyt replied:
Updated text in 1f1b977e

cthoyt committed:
Update docs

cthoyt committed:
Update output path

cthoyt committed:
Update models.py

cthoyt committed:
Merge branch 'master' into frequency-base-lines

cthoyt committed:
Merge branch 'master' into frequency-base-lines

cthoyt committed:
Update labels in plots

cthoyt committed:
Update models.py

cthoyt committed:
Make checks more explicit

cthoyt committed:
Update __init__.py

cthoyt committed:
Update test_models.py

cthoyt committed:
Merge branch 'master' into frequency-base-lines

mberr committed:
remove soft inverse triples baseline

mberr committed:
Re-generate readme

trigger ci

mberr committed:
cleanup experiment.py

trigger ci

mberr committed:
resolve todo

trigger ci

mberr committed:
remove left-over method

trigger ci

mberr commented on src/pykeen/models/__init__.py:
```diff
@@ -93,10 +97,14 @@
         ERModel,
         LiteralModel,
+        # baseline models behave differently
+        MarginalDistributionBaseline,
```
@cthoyt I made them unavailable via model_resolver, since they cannot be used, e.g., with the pipeline,

mberr commented on src/pykeen/models/baseline/experiment.py:
```diff
@@ -0,0 +1,230 @@
+# -*- coding: utf-8 -*-
```
@cthoyt should this be part of the main pykeen repo? Or should we move it to another repo (within the pykeen org)

cthoyt replied:
We can make an independent repo for it just like some of the other side projects. Let's keep it here until everything is done, though! It's nice to have for testing purposes

cthoyt replied:
See https://github.com/pykeen/nonparametric-baseline-benchmark

cthoyt commented on src/pykeen/models/baseline/experiment.py:
```diff
@@ -0,0 +1,230 @@
+        datasets = datasets[:1 + datasets.index(dataset_resolver.lookup('FB15k'))]
+    model_settings: List[Tuple[Type[EvaluationOnlyModel], Mapping[str, Any]]] = [
+        (MarginalDistributionBaseline, dict(entity_margin=True, relation_margin=True)),
+        # FIXME entity_margin=True, relation_margin=False throws an error
```
@mberr check this fixme

mberr replied:
bcf76ed

cthoyt commented on tests/test_baselines.py:
```diff
@@ -0,0 +1,46 @@
+        kwargs["triples_factory"] = self.factory
+        return kwargs
+
+    def test_score_t(self):
```
@mberr add tests for all combinations of entity_margin=False/True and relation_margin=False/True

mberr replied:
15f91a8

mberr committed:
Add tests for other combinations

mberr committed:
fix marginal_score for entity only margin

trigger ci

mberr committed:
fix mypy

trigger ci

mberr commented on src/pykeen/models/baseline/utils.py:
```diff
@@ -45,17 +45,17 @@ def marginal_score(
         scores = e_score.multiply(r_score)
         scores = sklearn_normalize(scores, norm="l1", axis=1)
+    else:
+        raise AssertionError  # for mypy
```
@cthoyt if you have a nicer solution, please go ahead :slightly_smiling_face:

cthoyt replied:
Nope this is good enough

cthoyt committed:
Remove experiment script

Trigger CI

Moved over to https://github.com/pykeen/nonparametric-baseline-benchmark

cthoyt committed:
Update __init__.py

Trigger CI
